File: StaticSharp/BasicComponents/Modifier/BaseModifier.cs
﻿using ImageMagick;
using StaticSharp.Gears;
using StaticSharp.Html;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;

namespace StaticSharp {

    namespace Js {
        public interface BaseModifier : Hierarchical {

            public new BaseModifier Parent { get; }
            public Color BackgroundColor { get; }
            public Color HierarchyBackgroundColor { get; }
            public Color ForegroundColor { get; }
            public Color HierarchyForegroundColor { get; }


            public bool Hover { get; }
            public bool Selectable { get; }

            public double Visibility { get; }

            public double Radius { get; }
            public double RadiusTopLeft { get; }
            public double RadiusTopRight { get; }
            public double RadiusBottomLeft { get; }
            public double RadiusBottomRight { get; }

        }
    }

    namespace Gears {
        public class BaseModifierBindings<FinalJs> : HierarchicalBindings<FinalJs> {
            public Binding<Color> BackgroundColor { set { Apply(value); } }
            public Binding<Color> ForegroundColor { set { Apply(value); } }
            public Binding<Color> TextDecorationColor { set { Apply(value); } }

            public Binding<bool> Selectable { set { Apply(value); } }
            public Binding<double> Visibility { set { Apply(value); } }            
            public Binding<double> Radius { set { Apply(value); } }
            public Binding<double> RadiusTopLeft { set { Apply(value); } }
            public Binding<double> RadiusTopRight { set { Apply(value); } }
            public Binding<double> RadiusBottomLeft { set { Apply(value); } }
            public Binding<double> RadiusBottomRight { set { Apply(value); } }

        }
    }

    namespace Gears {

        [Mix(typeof(BaseModifierBindings<Js.BaseModifier>))]

        [RelatedScript("../../CrossplatformLibrary/Color/Color")]
        [ConstructorJs]
        public abstract partial class BaseModifier: Hierarchical {

            public string? ExternalLink { get; set; }
            public Tree.Node? InternalLink { get; set; }
            public bool OpenLinksInANewTab { get; set; }
            
            public FontFamilyGenome[]? CodeFontFamilies { get; set; } = null;


            private FontFamilies? fontFamilies = null;
            public FontFamilies FontFamilies {
                get {
                    if (fontFamilies == null)
                        fontFamilies = new();
                    return fontFamilies;
                }
                set {
                    fontFamilies = value;
                }
            }



            public FontWeight? Weight { get; set; } = null;
            public bool? Italic { get; set; } = null;





            //public string? Url = null;
            public string? Tooltip = null;

            public double? LineHeight = null;//line-height
            public double? LetterSpacing = null;//letter-spacing


            protected BaseModifier(Hierarchical other, int callerLineNumber, string callerFilePath) : base(other, callerLineNumber, callerFilePath) {}
            public BaseModifier(int callerLineNumber, string callerFilePath) : base(callerLineNumber, callerFilePath) { }


            public string? GetUrl(Context context) {
                if (InternalLink != null) {
                    var url = context.NodeToUrlRelativeToCurrentNode(InternalLink);
                    return url.ToString();
                } else {
                    return ExternalLink;
                }
            }

            public override Tag GenerateHtml(Context context) {
                var url = GetUrl(context);
                if (url == null) {
                    return base.GenerateHtml(context);
                } else {
                    return new Tag("a") {
                        ["href"] = url,
                        ["target"] = OpenLinksInANewTab ? "_blank" : "_self",
                        Style = {
                            ["display"] = "contents",
                        },
                        Children = {
                            base.GenerateHtml(context)
                        }
                    };
                }
            }



            protected override Context ModifyContext(Context context) {
                if (fontFamilies != null) {
                    context.FontFamilies = fontFamilies;
                }
                if (CodeFontFamilies != null) {
                    context.CodeFontFamilies = CodeFontFamilies;
                }

                if (Weight != null) {
                    context.FontWeight = Weight.Value;
                }

                if (Italic != null) {
                    context.ItalicFont = Italic.Value;
                }


                return context;
            }

            protected override void ModifyHtml(Context context, Tag elementTag) {
                //protected void ModifyTag(Tag tag) {
                /*if (Url != null) {
                    tag.Name = "a";
                    tag["href"] = Url;
                }*/

                if (Tooltip != null) {
                    elementTag["title"] = Tooltip;
                }

                if (LineHeight != null) {
                    elementTag.Style["line-height"] = LineHeight;
                }

                if (LetterSpacing != null){
                    elementTag.Style["letter-spacing"] = LetterSpacing+"em";
                }

                if (fontFamilies != null) {
                    elementTag.Style["font-family"] = string.Join(',', fontFamilies.Select(x => x.Name));
                }

                if (Weight != null) {
                    elementTag.Style["font-weight"] = (int)Weight.Value;
                }

                if (Italic != null) {
                    elementTag.Style["font-style"] = Italic.Value ? "italic" : "normal";
                }


                base.ModifyHtml(context, elementTag);
            }

        }
    }
}

File: StaticSharp/BasicComponents/Modifier/BaseModifier.js
StaticSharpClass("StaticSharp.BaseModifier", (element) => {

    StaticSharp.Hierarchical(element)


    //TODO: fix file protocol
    /*if (element.parentElement.tagName == "A") {
        if (window.location.protocol == "file:") {
            let a = element.parentElement
            let href = a.getAttribute("href")
            let AbsoluteUrlRegExp = new RegExp('^(?:[a-z+]+:)?//', 'i')
            if (!AbsoluteUrlRegExp.test(href)) {
                a.setAttribute("href", href + ".html")
            }
        }        
    }*/
    


    //var extension = (window.location.protocol == "file:") ? ".html" : ""
    

    //window.location.replace(matchLanguage([{{ languages }}]) + extension)

    /**type */

    let baseAs = element.as
    element.as = function (typeName) {
        

        let result = baseAs(typeName)
        if (result != undefined)
            return result

        
        

        if (element.Modifiers != undefined) {
            /*for (let i of element.Modifiers) {
                console.log("as", i.is)
            }*/



            let oldAs = element.as
            element.as = () => undefined
            result = element.Modifiers.find(x => x.is(typeName))
            element.as = oldAs
        }
        
        return result
    }


    //element.Modifiers = undefined,

    element.Reactive = {
        BackgroundColor: undefined,
        HierarchyBackgroundColor: () => element.BackgroundColor || element.Parent.HierarchyBackgroundColor,

        ForegroundColor: () => { 
            if (element.BackgroundColor != undefined)
                return element.BackgroundColor.ContrastColor()
            else
                return undefined
        },
        HierarchyForegroundColor: () => {
            let result = element.ForegroundColor
            if (result) return result
            let parent = element.Parent
            if (parent) {
                result = parent.HierarchyForegroundColor
                if (result) return result
            }
            return new Color(0)
        },

        TextDecorationColor: () => {
            if (element.ForegroundColor != undefined)
                return element.ForegroundColor
            else
                return undefined
        },

        Visibility: 1
    }

    new Reaction(() => {
        let visibility = element.Visibility
        if (visibility == 0) {
            element.style.visibility = "hidden"
            element.style.opacity = ""
        } else if (visibility == 1) {
            element.style.visibility = ""
            element.style.opacity = ""
        } else {
            element.style.opacity = visibility
            element.style.visibility = ""
        }        
    })
    

    new Reaction(() => {
        element.style.backgroundColor = ToCssValue(element.BackgroundColor)
    })

    new Reaction(() => {
        element.style.color = ToCssValue(element.ForegroundColor)
    })

    new Reaction(() => {
        element.style.textDecorationColor = ToCssValue(element.TextDecorationColor)
    })
})
